lib/network/level-net/index.js

```diff
@@ -0,0 +1,343 @@
+        };
+    }
+    for (const remoteCall of Object.keys(miscSyncCommands).concat(
+        Object.keys(miscAsyncCommands))) {
+        dbClient[remoteCall] = function onCall(...rpcArgs) {
+            const cb = rpcArgs.pop();
+            const args = { rpcArgs };
+            this.miscClient.callTimeout(remoteCall, args, cb,
+                                        this.callTimeoutMs);
+        };
+    }
+
+    /**
+     * @brief return a handle to a sublevel database
+     *
+     * @note this function has no side-effect on the db, it just
+     * returns a handle properly configured to access the sublevel db
+     * from the client.
+     *
+     * @param {String} subName name of sublevel
+     * @return {Object} a handle to the sublevel database that has the
+     * same API as its parent
+     */
+    dbClient.openSub = function openSub(subName) {
+        const subLevel = {};
+        Object.assign(subLevel, this);
+        // maintain path as a list of nested sublevels
+        subLevel.path = subLevel.path.slice();
+        subLevel.path.push(subName);
+        return subLevel;
+    };
+
+    return dbClient;
+};
+
+/**
+ * @brief create a server object that serves remote LevelDB requests
+ * through socket.io events
+ *
+ * @param {Object} db The root LevelDB database object to expose to
+ * remote clients
+ * @param {Object} options options object
+ * @param {Object} options.logger logger object
+ * @param {Number} [options.streamMaxPendingAck] max number of
+ *   in-flight output stream packets sent to the server without an ack
+ *   received yet
+ * @param {Number} [options.streamAckTimeoutMs] timeout for receiving
+ *   an ack after an output stream packet is sent to the server
+ * @return {Object} a server object, not yet listening on a TCP port
+ * (you must call listen(port) on the returned object)
+ */
+module.exports.createServer = function LevelServer(db, options) {
+    assert(options.logger);
+
+    const httpServer = http.createServer();
+    const ioServer = io(httpServer);
+    const log = options.logger;
+
+    ioServer.initMetadataService = function initMetadataService(baseNs = '') {
+        assert(baseNs === '' || baseNs.startsWith('/'));
+
+        // all metadata operations executed by leveldb go through this
+        // namespace, other operations (ping etc.) go through /misc.
+        const mdSock = this.of(`${baseNs}/metadata`);
+        mdSock.on('connection', conn => {
+            /**
+             * lookup a sublevel db given by the @a path array from
+             * the main db handle.
+             *
+             * @param {String []} path path to the sublevel, as a
+             * piecewise array of sub-levels
+             * @return {Object} the handle to the sublevel
+             */
+            function lookupSubLevel(path) {
+                let subDb = db;
+                path.forEach(pathItem => {
+                    subDb = subDb.sublevel(pathItem);
```

Comment:
This sounds like you have a tree of sub-level dbs that you have to search for each using its absolute path. If what I understood is correct then do you think it's necessary to map a flat structure (buckets) to this hierarchical structure (even though it looks like there are only two levels: namespace and bucket)?